### PR TITLE
Mp jwt12 encrypt fat 2

### PR DESCRIPTION
**Pull request by c00crane**
Deliver FAT tests for JWE support in MP Jwt.

Tests were added to validate behavior with keyManagementKeyAlias configured in server.xml or mp.jwt.decrypt.key.location in microprofile-config.properties within an app, or as environment variables or system properties.
Tests use valid and invalid values.
Tests use tokens that either match what is configured in mpJwt or don't match to show that we can consume the token properly as well as detect problems with the token and block consumption.

### CODE REVIEW

**Comment by c00crane** [comment]
#13575

**Comment by c00crane** [comment]
#build

File: dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/sharedTests/MPJwt12MPConfigTests.java [hunk 1]
```diff
@@ -519,58 +650,191 @@ public Expectations setBadAudiencesExpectations(LibertyServer server) throws Exc
...
+     * Sets expectations to check when the keyManagementKeyAlias is not set
+     *
+     * @param server - server whose logs will be searched
+     * @param extraMsgs - the tai drives the code down different paths depending on if it finds config info in server.xml - if it finds config settings, we'll get 2 extra messages.
+     * @return Expectations - built expectations
+     * @throws Exception
+     */
+    public Expectations setEncryptMismatchKeyTypeExpectations(LibertyServer server, boolean extraMsgs) throws Exception {
+        Expectations expectations = setAllBadEncryptExpectations(server, extraMsgs);
+        expectations.addExpectation(new ServerMessageExpectation(server, "java.lang.ClassCastException", "Messagelog did not contain an exception indicating a classcast exception due to the key type."));
```
Comment (ayoho) [inline dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/sharedTests/MPJwt12MPConfigTests.java:708]: I might be interested in double checking this one. Do you know whether the `ClassCastException` is coming from our code or from jose4j? I wonder if there might be something cleaner we could do, depending on what this actually ends up looking like.

**Review by ayoho — APPROVED** [review]

**Comment by c00crane** [comment]
#run-libby-bot

File: dev/com.ibm.ws.security.mp.jwt.1.2_fat/bnd.bnd [hunk 2]
```diff
@@ -11,6 +11,14 @@
 -include= ~../cnf/resources/bnd/bundle.props
 bVersion=1.0
 
+Import-Package: \
+	!*.internal.*,\
+	!com.ibm.ws.kernel.boot.cmdline,\
+	com.google.gson.*:resolution:=optional,\
+	org.jose4j.*;version="[0.6,5)", \
+	org.apache.commons.codec.binary,\
+	*
```
[the remark this answers is not in the document]
Comment (c00crane) [reply dev/com.ibm.ws.security.mp.jwt.1.2_fat/bnd.bnd:20]: removed

File: dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/configInAppTests/MPJwt12MPConfigInApp_Tests.java [hunk 3]
```diff
@@ -594,8 +607,481 @@ public void MPJwt12MPConfigInApp_GoodNonDefaultKeyAndAlgorithmInMPConfig_UnderWe
...
+        String jwtToken = builderHelpers
+                        .buildJWETokenWithAltHeader(JwtKeyTools.getPublicKeyFromPem(JwtKeyTools.getComplexPublicKeyForSigAlg(jwtBuilderServer, MpJwt12FatConstants.SIGALG_RS256)),
+                                                    "notJOSE", "jwt");
+
+        // The test code generates a token that uses HS256 - if we get far enough to fail on that, we haven't failed checking the JWE Type :)
+        useToken(jwtToken,
+                 buildAppUrl(resourceServer, MpJwt12FatConstants.GOOD_RELATIVE_DECRYPT_KEY_RS256_IN_CONFIG_IN_META_INF_ROOT_CONTEXT,
+                             MpJwt12FatConstants.MP_CONFIG_IN_META_INF_TREE_APP),
+                 MpJwt12FatConstants.MP_CONFIG_IN_META_INF_TREE_APP, MpJwt12FatConstants.AUTHORIZATION,
+                 MpJwt12FatConstants.TOKEN_TYPE_BEARER, setBadCertExpectations(resourceServer, KeyMismatch));
```
[the remark this answers is not in the document]
Comment (c00crane) [reply dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/configInAppTests/MPJwt12MPConfigInApp_Tests.java:1000]: same comment... updated the comments

File: dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/sharedTests/GenericEnvVarsAndSystemPropertiesTests.java [hunk 4]
```diff
@@ -68,23 +79,59 @@ public static void commonMpJwt12Setup(LibertyServer requestedServer, String conf
...
+     * Caller will set up mp config properties expecting rs256 encryption and signing in either env vars or system properties
+     * We can't specify values other than the defaults for these 2 attributes in the builder config, so, we need to build our own token
+     *
+     * @param keyMgmtAlg - requested key management key alg to use when building the token
+     * @param contentEncryptAlg - requested content encryption algorithm to use when building the token
+     * @throws Exception
+     */
+    public void genericDecryptOtherKeyMgmtAlgOrOtherContentEncryptAlg(String keyMgmtAlg, String contentEncryptAlg) throws Exception {
+
+        List<NameValuePair> extraClaims = new ArrayList<NameValuePair>();
```
[the remark this answers is not in the document]
Comment (c00crane) [reply dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/sharedTests/GenericEnvVarsAndSystemPropertiesTests.java:121]: Using what called method expects passed - if I use a map, I'd probably have to re-write called method and other callers of that method.

**Comment by c00crane** [comment]
#build
#run-libby-bot

**Comment by c00crane** [comment]
#run-libby-bot

File: dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/MPJwt12ConfigUsingBuilderTests.java [hunk 5]
```diff
@@ -293,71 +304,293 @@ public void MPJwt12ConfigUsingBuilderTests_Header_Cookie_passTokenInAuthHeaderAn
...
+                                                    "notJOSE", "jwt");
+// for debug       new JwtTokenForTest(jwtToken, JwtKeyTools.getComplexPrivateKeyForSigAlg(jwtBuilderServer, MpJwt12FatConstants.SIGALG_RS256));
+        Log.info(thisClass, _testName, "Funky token: " + jwtToken);
+        for (TestApps app : setTestAppArray(resourceServer)) {
+            WebClient webClient = actions.createWebClient();
+
+            Page response = actions.invokeUrlWithBearerToken(_testName, webClient, app.getUrl(), jwtToken);
+
+            // The test code generates a token that uses HS256 - if we get far enough to fail on that, we haven't failed checking the JWE Type :)
+            validationUtils.validateResult(response, setBadCertExpectations(resourceServer, KeyMismatch));
```
[the remark this answers is not in the document]
Comment (c00crane) [reply dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/MPJwt12ConfigUsingBuilderTests.java]: The test "builder" is building a token that is encrypted with RS256, but signed with HS256.  The mpJwt config is expecting RS256 - hence the key mismatch failure on the signature.  It should never get a message saying that the payload isn't JWS - we're expecting JWE and we're sending JWE - the header of the JWE just has a different "typ" value - which we're trying to show that we don't care/check.
I fixed the comment before the call to build the token as well as the comment on the check.

**Comment by c00crane** [comment]
#build
#run-libby-bot

**Pull request merged**